### PR TITLE
Introduce PlutoContext; drop global variable PlutoOptions

### PR DESCRIPTION
**Pull request by bondhugula**
- introduce PlutoContext and drop the global var PlutoOptions;
     options can now be obtained from prog->context->options;
     besides PlutoProg, PlutoConstraints and PlutoMatrix also hold
     a pointer to the context.

   - libpluto is now thread-safe

Signed-off-by: Uday Bondhugula <uday@polymagelabs.com>

### CODE REVIEW

**Comment by bondhugula** [comment]
Update framework-dfp.c as well. @aravind343

File: src/plutolib.c [hunk 1]
```diff
@@ -139,14 +140,19 @@ static isl_stat extract_basic_set(__isl_take isl_basic_set *bset, void *user) {
   return isl_stat_ok;
 }
 
+struct pluto_stmt_context_info {
+  Stmt **stmts;
+  PlutoContext *context;
+};
+
 /* Used by libpluto interface. Expects that stmts[idx] have been initialized to
  * NULL.  */
```
Comment (aravind343) [inline src/plutolib.c]: Doc style /// comment.

File: src/math_support.c [hunk 2]
```diff
@@ -271,26 +271,20 @@ void pluto_matrix_set(PlutoMatrix *mat, int val) {
 }
 
 /* Return an identity matrix of size: size x size */
```
Comment (aravind343) [inline src/math_support.c]: Doc style comment.

File: src/osl_pluto.c [hunk 3]
```diff
@@ -734,7 +738,7 @@ static PlutoConstraints *osl_relation_to_pluto_constraints(osl_relation_p rln) {
 }
 
 /* Read statement info from openscop structures (nvar: max domain dim) */
-static Stmt **osl_to_pluto_stmts(const osl_scop_p scop) {
```
Comment (aravind343) [inline src/osl_pluto.c]: Doc style /// comment.

File: src/isl_support.c [hunk 4]
```diff
@@ -25,8 +25,11 @@
 #include <stdlib.h>
 #include <string.h>
 
+#include "isl_support.h"
+
```
[the remark this answers is not in the document]
Comment (bondhugula) [reply src/isl_support.c]: The header file corresponding to that file always comes first - the rest come in alphabetical order.

File: src/osl_pluto.c [hunk 5]
```diff
@@ -1586,10 +1599,11 @@ static Dep **deps_read(osl_dependence_p candlDeps, PlutoProg *prog) {
  * representation of the program sufficient to be used throughout Pluto.
  * PlutoProg also includes dependences; so candl is run here.
  */
```
[the remark this answers is not in the document]
Comment (bondhugula) [reply src/osl_pluto.c]: Done.

File: src/osl_pluto.c [hunk 6]
```diff
@@ -1481,7 +1494,7 @@ static Dep **deps_read(osl_dependence_p candlDeps, PlutoProg *prog) {
     dep->dest = candl_dep->label_target;
 
     // candl_matrix_print(stdout, candl_dep->domain);
```
[the remark this answers is not in the document]
Comment (bondhugula) [reply src/osl_pluto.c]: Done.

File: test.sh.in [hunk 7]
```diff
@@ -71,7 +71,7 @@ for file in $TESTS; do
 done
 
 # Test per cc objective
-printf '%-50s ' test/test-per-cc-obj.c
+printf '%-50s ' ./test/test-per-cc-obj.c
```
[the remark this answers is not in the document]
Comment (bondhugula) [reply test.sh.in]: Thanks.

File: src/plutolib.c (hunk 1 above)
Comment (bondhugula) [reply src/plutolib.c]: This is really not documentation but a comment on the working of a local function. So just "// ".

File: src/osl_pluto.c (hunk 3 above)
Comment (bondhugula) [reply src/osl_pluto.c]: Done.

**Pull request merged**